tools/degreeworks-scraper/src/components/DegreeworksClient.ts

```diff
@@ -0,0 +1,131 @@
+import fetch from "cross-fetch";
+
+import type { Block, DWAuditResponse, DWMappingResponse } from "../types";
+
+export class DegreeworksClient {
+  private static readonly API_URL = "https://reg.uci.edu/RespDashboard/api";
+  private static readonly AUDIT_URL = `${DegreeworksClient.API_URL}/audit`;
+  private catalogYear: string = "";
+
+  private constructor(
+    private readonly studentId: string,
+    private readonly headers: HeadersInit,
+    private readonly delay: number,
+  ) {}
+
+  static async new(
+    studentId: string,
+    headers: HeadersInit,
+    delay: number = 1000,
+  ): Promise<DegreeworksClient> {
+    const dw = new DegreeworksClient(studentId, headers, delay);
+    /**
+     * Depending on when we are scraping, the catalog year may be the academic year that
+     * started the previous calendar year, or the one that will start this calendar year.
+     *
+     * We determine the catalog year by seeing if we can fetch the major data for the
+     * B.S. in Computer Science for the latter. If it is available, then we use that
+     * as the catalog year. Otherwise, we use the former.
+     */
+    const currentYear = new Date().getUTCFullYear();
+    dw.catalogYear = `${currentYear}${currentYear + 1}`;
+    if (!(await dw.getMajorAudit("BS", "U", "201"))) {
+      dw.catalogYear = `${currentYear - 1}${currentYear}`;
+    }
```

Comment:
Could this be more descriptive? Directly using promises can be opaque.
```suggestion
    const currentYear = new Date().getUTCFullYear();
    const dataThisYear = await dw.getMajorAudit("BS", "U", "201")
    dw.catalogYear = dataThisYear != null 
       ? `${currentYear}${currentYear + 1}`
       : `${currentYear - 1}${currentYear}`
```